lib/promscrape/scrapework.go

```diff
@@ -527,51 +526,58 @@ func (sw *scrapeWork) processDataOneShot(scrapeTimestamp, realTimestamp int64, b
 }
 
 func (sw *scrapeWork) processDataInStreamMode(scrapeTimestamp, realTimestamp int64, body *bytesutil.ByteBuffer, scrapeDurationSeconds float64) error {
-	samplesScraped := 0
-	samplesPostRelabeling := 0
-	wc := writeRequestCtxPool.Get(sw.prevLabelsLen)
+	var samplesScraped atomic.Int64
+	var samplesPostRelabeling atomic.Int64
+	var samplesDroppedTotal atomic.Int64
 
 	lastScrape := sw.loadLastScrape()
 	bodyString := bytesutil.ToUnsafeString(body.B)
 	areIdenticalSeries := sw.areIdenticalSeries(lastScrape, bodyString)
-	samplesDropped := 0
 
 	r := body.NewReader()
-	var mu sync.Mutex
 	err := stream.Parse(r, scrapeTimestamp, "", false, func(rows []parser.Row) error {
-		mu.Lock()
-		defer mu.Unlock()
+		var err error
+
+		wc := writeRequestCtxPool.Get(sw.prevLabelsLen)
+		defer func() {
+			wc.resetNoRows()
+			writeRequestCtxPool.Put(wc)
+		}()
+		samplesScraped.Add(int64(len(rows)))
 
-		samplesScraped += len(rows)
 		for i := range rows {
 			sw.addRowToTimeseries(wc, &rows[i], scrapeTimestamp, true)
 		}
-		samplesPostRelabeling += len(wc.writeRequest.Timeseries)
-		if sw.Config.SampleLimit > 0 && samplesPostRelabeling > sw.Config.SampleLimit {
+		newSamplesPostRelabeling := samplesPostRelabeling.Add(int64(len(wc.writeRequest.Timeseries)))
+		if sw.Config.SampleLimit > 0 && int(newSamplesPostRelabeling) > sw.Config.SampleLimit {
 			wc.resetNoRows()
 			scrapesSkippedBySampleLimit.Inc()
-			return fmt.Errorf("the response from %q exceeds sample_limit=%d; "+
+			err = fmt.Errorf("the response from %q exceeds sample_limit=%d; "+
 				"either reduce the sample count for the target or increase sample_limit", sw.Config.ScrapeURL, sw.Config.SampleLimit)
 		}
-		if sw.seriesLimitExceeded || !areIdenticalSeries {
-			samplesDropped += sw.applySeriesLimit(wc)
+
+		if sw.seriesLimitExceeded.Load() || !areIdenticalSeries {
+			if samplesDropped := sw.applySeriesLimit(wc); samplesDropped > 0 {
+				samplesDroppedTotal.Add(int64(samplesDropped))
+			}
 		}
 
 		// Push the collected rows to sw before returning from the callback, since they cannot be held
 		// after returning from the callback - this will result in data race.
 		// See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/825#issuecomment-723198247
 		sw.pushData(sw.Config.AuthToken, &wc.writeRequest)
 		wc.resetNoRows()
-		return nil
+		return err
 	}, sw.logError)
 
-	scrapedSamples.Update(float64(samplesScraped))
+	scrapedSamples.Update(float64(samplesScraped.Load()))
 	up := 1
 	if err != nil {
 		// Mark the scrape as failed even if it already read and pushed some samples
 		// to remote storage. This makes the logic compatible with Prometheus.
 		up = 0
 		scrapesFailed.Inc()
+		bodyString = ""
```

Comment:
align stream behavior with a one-shot one. It sets `scrapeResponseSize` metric to zero on error.